openkaito/evaluation/evaluator.py

```diff
@@ -219,15 +274,157 @@ def llm_ranking_evaluation(self, query_string, size, docs, retries=3):
                 raise ValueError(
                     f"Length of ranking {len(ranking)} does not match input docs length {len(docs)}"
                 )
-            ranking_score = ndcg_score(ranking, size)
-            bt.logging.info(f"LLM Ranking score: {ranking_score}")
-            return ranking_score
+            # ranking_score = ndcg_score(ranking, size)
+            # bt.logging.info(f"LLM Ranking score: {ranking_score}")
+            # return ranking_score
+            return ranking
         except Exception as e:
             bt.logging.error(f"Error while parsing LLM result: {e}, retrying...")
             if retries > 0:
-                return self.llm_ranking_evaluation(query_string, docs, retries - 1)
+                return self.llm_keyword_ranking_evaluation(
+                    query_string, docs, retries - 1
+                )
             else:
                 bt.logging.error(
-                    f"Failed to parse LLM result after retrying. Returning 0."
+                    f"Failed to parse LLM result after retrying. Returning [0]."
                 )
+            return [0]
+
+    def llm_structured_search_ranking_evaluation(self, query_string, docs, retries=3):
+        """
+        This function evaluates the structured search ranking of the documents using the LLM.
+        Here, the `query_string` may contains `AND`/`OR` operators in the query.
+        """
+        try:
+            newline = "\n"
+            prompt_docs = "\n\n".join(
+                [
+                    f"ItemId: {i}\nTime: {doc['created_at'].split('T')[0]}\nText: {doc['text'][:1000].replace(newline, '  ')}"
+                    for i, doc in enumerate(docs)
+                ]
+            )
+            bt.logging.debug(
+                f"Querying LLM of {query_string} with docs:\n" + prompt_docs
+            )
+            output = self.llm_client.chat.completions.create(
+                model="gpt-3.5-turbo",
+                response_format={"type": "json_object"},
+                messages=[
+                    {
+                        "role": "system",
+                        "content": """Below are the metrics and definations: 
+Outdated: Time-sensitive information that is no longer current or relevant.
+Off topic: Superficial content lacking depth and comprehensive insights.
+Somewhat Relevant: Offers partial insight but lacks depth and comprehensive coverage.
+Relevant: Comprehensive, insightful content suitable for informed decision-making.""",
+                    },
+                    {
+                        "role": "system",
+                        "content": """You will be given a list of documents with id and you have to rate them based on the relevance to the structued query. 
+The structured query can contain AND/OR operators. 
+AND operator: The document should contain all the keywords in the query.
```

Comment:
Should we make this more semantic? Basically AND means the doc contains meaningful discussions for both A and B, OR means contains meaningful discussions for either A OR B. I think that was reflected in the in-context example, but wondering if we should also focus less on keyword match here.

Reply:
yes, agree!